design-documents/bit-precise-types.rst

```diff
@@ -0,0 +1,506 @@
+bit-precise integral types defined in C2x.  These are ``_BitInt(N)`` and
+``unsigned _BitInt(N)``.  These are defined for integral ``N`` and each ``N`` is
+a different type.
+
+The proposal for these types can be found in following link.
+https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2763.pdf
+
+As the rationale in that proposal mentioned, some applications have uses for a
+specific bit-width type.  In the case of writing C code which can be used to
+determine FPGA hardware these specific bit-width types can lead to large
+performance and space savings.
+
+From the perspective of the Arm ABI we have some trade-offs and decisions to
+make:
+
+- We need to choose a representation for these objects in registers.
+- We need to choose a representation, size and alignment of these objects in memory.
+
+The main trade-offs we have identified in this case are:
+
+- Performance of different C-level operations.
+- Whether certain hardware-level atomic operations are possible.
+- Size cost of storing values in memory.
+- General familiarity of programmers with the representation.
+
+Since this is a new type there is large uncertainty on how it will be used by
+programmers in the future.  Decisions we make here may also influence future
+usage.  Nonetheless we must make trade-off decisions with this uncertainty.  The
+below attempts to analyze possible use-cases to make our best guess as to how
+these types may be used when targeting Arm CPU's.
+
+
+Use-cases known of so far
+-------------------------
+
+There seem to be two different regimes for these types.  The "small" regime
+where bit-precise types could be stored in a single general-purpose register,
+and the "large" regime where bit-precise types must span multiple
+general-purpose registers.
+
+Here we discuss the use-cases for bit-precise integer types that we have
+identified or been alerted to so far.
+
+
+C code to describe FPGA behavior
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A major motivating use-case for this new type is to aid writing C code which
+describes the desired behavior of an FPGA.  Without the availability of the new
+``_BitInt`` type such C code would semantically have much wider types than
+necessary when performing operations, especially given that operations on small
+integral types promote their operands to ``int``.
+
+If these wider than necessary operations end up in the FPGA they would use many
+more many more logic gates than necessary.  Using ``_BitInt`` allows programmers
+to write code which directly expresses what is needed.  This can ensure the FPGA
+description generated saves space and has better performance.
+
+The notable thing about this use-case is that though the C code may be run on an
+Arm architecture (e.g. for testing), the most critical use is when transferred
+to an FPGA (i.e. not an Arm architecture).
+
+That said, if the operation that this FPGA performs becomes popular there may be
```

Comment:
```suggestion
However, if the operation that this FPGA performs becomes popular there may be
```

Reply:
Would like to not make this change -- I believe the change puts much more emphasis on the second point than it did before, and that's not emphasis that I believe should be there (at least that's what "However" sounds like to me).

Is this suggestion in order to make things easier to read for non-native speakers (as the suggestions to avoid latin contractions), or is it for some other reason?
(Would make the change if the point is ease-of-understanding for non-native speakers, reserve judgement for other motivations ;-).